File: bindings/zig/src/drtrace_test.zig
//! Standalone smoke test for the DynamoRIO native-trace wrapper (drtrace.zig).
//!
//! The Zig analogue of `bindings/python/tests/test_drtrace.py`. Built as a
//! separate `drtrace-test` step (NOT part of `zig build test`) because it must
//! NOT link `libasmtest_drapp` at build time — the lib needs DynamoRIO and may
//! be absent. It dlopen()s the lib at runtime and **self-skips** cleanly: if
//! `drtrace.available()` is false it prints `SKIP: ...` and exits 0.
//!
//! A `main` (not a `test` block) so it doubles as `zig run src/drtrace_test.zig`
//! and as an `addExecutable` run artifact. Uses explicit `if (!cond) return
//! error.Failed` checks rather than `std.debug.assert` (assert is compiled out
//! in ReleaseFast) so the assertions keep their teeth in any optimize mode.
const std = @import("std");
const drtrace = @import("drtrace.zig");

// mov rax,rdi; add rax,rsi; cmp rax,100; jle +3; dec rax; ret  (two basic blocks)
const ROUTINE = [_]u8{
    0x48, 0x89, 0xF8, 0x48, 0x01, 0xF0, 0x48, 0x3D, 0x64, 0x00,
    0x00, 0x00, 0x7E, 0x03, 0x48, 0xFF, 0xC8, 0xC3,
};

const Failure = error{Failed};

fn check(cond: bool, comptime msg: []const u8) Failure!void {
    if (!cond) {
        std.debug.print("FAIL: {s}\n", .{msg});
        return Failure.Failed;
    }
}

// `region` body helpers: invoke the code and stash the result so the caller can
// read it after the scoped begin/end closes.
var g_result: c_long = 0;
fn callBody(code: *const drtrace.NativeCode, a: c_long, b: c_long) void {
    g_result = code.call(a, b);
}

pub fn main() !void {
    if (!drtrace.available()) {
        std.debug.print("SKIP: DynamoRIO native-trace tier unavailable (self-skip)\n", .{});
        return; // exit 0
    }

    drtrace.initializeDefault() catch |e| {
        // dr_init/start can fail even when "available" if the client/env is
        // misconfigured; treat as a skip like the Python fixture does.
        std.debug.print("SKIP: dr_init/start failed: {s}\n", .{@errorName(e)});
        return;
    };
    defer drtrace.shutdown();

    // ---- block-coverage + accumulation ---- //
    var code = try drtrace.NativeCode.fromBytes(&ROUTINE);
    var tr = try drtrace.NativeTrace.create(64, 0); // blocks=64, instructions=0
    try tr.register("add2", &code);

    tr.region("add2", .{ &code, @as(c_long, 20), @as(c_long, 22) }, callBody);
    try check(g_result == 42, "add2(20,22) == 42");
    try check(tr.covered(0), "entry block (offset 0) covered");

    const before = tr.blocksLen();
    tr.region("add2", .{ &code, @as(c_long, 60), @as(c_long, 60) }, callBody);
    try check(g_result == 119, "add2(60,60) == 119 (other block: 120>100 -> dec)");
    try check(tr.blocksLen() >= before, "blocks_len accumulates");
    try check(drtrace.markerError() == 0, "marker_error == 0 (markers balanced)");

    tr.unregister("add2");
    code.free();
    tr.free();

    // ---- instruction mode ---- //
    var code2 = try drtrace.NativeCode.fromBytes(&ROUTINE);
    var tr2 = try drtrace.NativeTrace.create(64, 64); // blocks=64, instructions=64
    try tr2.register("add2i", &code2);

    tr2.region("add2i", .{ &code2, @as(c_long, 1), @as(c_long, 2) }, callBody);
    try check(g_result == 3, "add2i(1,2) == 3");
    try check(tr2.insnsTotal() >= 4, "insns_total >= 4 (ordered instruction stream)");

    const alloc = std.heap.page_allocator;
    const insns = try tr2.insnOffsets(alloc);
    defer alloc.free(insns);
    try check(std.mem.eql(u64, insns, &[_]u64{ 0x0, 0x3, 0x6, 0xc, 0x11 }),
        "insn_offsets == [0x0,0x3,0x6,0xc,0x11]");

    const blocks = try tr2.blockOffsets(alloc);
    defer alloc.free(blocks);
    try check(std.mem.indexOfScalar(u64, blocks, 0) != null, "block_offsets contains 0");

    tr2.unregister("add2i");
    code2.free();
    tr2.free();

    // ---- symbol mode ---- //
    // Trace a named exported function by NAME, no begin/end markers — recording
    // is always on for the symbol's range. `asmtest_symbol_demo` lives in the
    // drapp lib and is resolvable by the client's symbol search.
    var tr3 = try drtrace.NativeTrace.create(64, 0); // blocks=64, instructions=0
    try tr3.registerSymbol("asmtest_symbol_demo", 256);

    try check(drtrace.symbolDemo(3, 4) == 10, "symbol_demo(3,4) == 10 (a*2+b)");
    try check(tr3.covered(0), "symbol entry block (offset 0) covered");
    try check(drtrace.markerError() == 0, "marker_error == 0 (no markers used)");

    tr3.unregister("asmtest_symbol_demo");
    tr3.free();

    std.debug.print("PASS\n", .{});
}

File: bindings/java/DrTraceTest.java
/*
 * DrTraceTest.java — standalone smoke test for the DynamoRIO native-trace
 * binding (DrTrace.java), mirroring bindings/python/tests/test_drtrace.py.
 *
 * Self-skips unless the tier is built AND DynamoRIO is resolvable — i.e. unless
 * ASMTEST_DRAPP_LIB / ASMTEST_DRCLIENT (and ASMTEST_DR_LIB or DYNAMORIO_HOME)
 * point at a built libasmtest_drapp + client on a DynamoRIO-capable Linux
 * x86-64 host. On a skip it prints "SKIP: ..." and exits 0.
 *
 * Asserts are NOT used (the `assert` keyword is off by default); failures throw
 * or System.exit nonzero explicitly. On success it prints "PASS".
 *
 * Compile with `--release 22`, run with
 * `--enable-native-access=ALL-UNNAMED`.
 */
public final class DrTraceTest {

    // mov rax,rdi; add rax,rsi; cmp rax,100; jle +3; dec rax; ret  (two basic blocks)
    private static final byte[] ROUTINE = {
        0x48, (byte) 0x89, (byte) 0xF8, 0x48, 0x01, (byte) 0xF0, 0x48, 0x3D,
        0x64, 0x00, 0x00, 0x00, 0x7E, 0x03, 0x48, (byte) 0xFF, (byte) 0xC8, (byte) 0xC3
    };

    private static void check(boolean cond, String msg) {
        if (!cond) throw new AssertionError(msg);
    }

    public static void main(String[] args) {
        if (!DrTrace.available()) {
            Throwable e = DrTrace.loadError();
            System.out.println("SKIP: DynamoRIO native-trace tier unavailable (self-skip)"
                + (e != null ? ": " + e : ""));
            System.exit(0);
        }

        DrTrace.initialize(); // null client → C falls back to $ASMTEST_DRCLIENT
        try {
            blockCoverageAndAccumulation();
            instructionMode();
            symbolMode();
        } finally {
            DrTrace.shutdown();
        }
        System.out.println("PASS");
    }

    // Mirrors test_block_coverage_and_accumulation.
    private static void blockCoverageAndAccumulation() {
        DrTrace.NativeCode code = DrTrace.NativeCode.fromBytes(ROUTINE);
        DrTrace.NativeTrace tr = DrTrace.NativeTrace.create(64, 0);
        tr.register("add2", code);

        long[] r = new long[1];
        tr.region("add2", () -> r[0] = code.call(20, 22));
        check(r[0] == 42, "call(20,22): got " + r[0] + ", want 42");
        check(tr.covered(0), "entry block (offset 0) expected covered");

        long before = tr.blocksLen();
        tr.region("add2", () -> r[0] = code.call(60, 60)); // 120 > 100 → dec → 119, other block
        check(r[0] == 119, "call(60,60): got " + r[0] + ", want 119");
        check(tr.blocksLen() >= before, "blocksLen did not grow: " + tr.blocksLen() + " < " + before);
        check(DrTrace.markerError() == 0, "markerError: " + DrTrace.markerError());

        tr.unregister("add2");
        code.free();
        tr.free();
    }

    // Mirrors test_instruction_mode.
    private static void instructionMode() {
        DrTrace.NativeCode code = DrTrace.NativeCode.fromBytes(ROUTINE);
        DrTrace.NativeTrace tr = DrTrace.NativeTrace.create(64, 64);
        tr.register("add2i", code);

        long[] r = new long[1];
        tr.region("add2i", () -> r[0] = code.call(1, 2));
        check(r[0] == 3, "call(1,2): got " + r[0] + ", want 3");
        check(tr.insnsTotal() >= 4, "insnsTotal: got " + tr.insnsTotal() + ", want >= 4");

        long[] insns = tr.insnOffsets();
        long[] wantInsns = {0x0, 0x3, 0x6, 0xc, 0x11};
        check(java.util.Arrays.equals(insns, wantInsns),
            "insnOffsets: got " + java.util.Arrays.toString(insns)
                + ", want " + java.util.Arrays.toString(wantInsns));

        long[] blocks = tr.blockOffsets();
        boolean hasZero = false;
        for (long b : blocks) if (b == 0) { hasZero = true; break; }
        check(hasZero, "blockOffsets: " + java.util.Arrays.toString(blocks) + " does not contain 0");

        tr.unregister("add2i");
        code.free();
        tr.free();
    }

    // Mirrors test_symbol_mode: trace an exported function by name, no markers.
    private static void symbolMode() {
        DrTrace.NativeTrace tr = DrTrace.NativeTrace.create(64, 0);
        tr.registerSymbol("asmtest_symbol_demo", 256);

        long r = DrTrace.symbolDemo(3, 4); // no region/markers — always-on recording
        check(r == 10, "symbolDemo(3,4): got " + r + ", want 10");
        check(tr.covered(0), "entry block (offset 0) expected covered");
        check(DrTrace.markerError() == 0, "markerError: " + DrTrace.markerError());

        tr.unregister("asmtest_symbol_demo");
        tr.free();
    }
}
